[torchgen] Add logic in annotation parser to accept alias set

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83501

Extending the current regex in `model.py` to support annotation alias set. See issue #83214.

Ideally we should have a full fledged lexer similar to `schema_type_parser.cpp`, since regex can be more and more difficult to read if we add more support to it.

Adding this to unblock this issue for now.

larryliu0820 committed:
[torchgen] Add logic in annotation parser to accept alias set

Extending the current regex in `model.py` to support annotation alias set. See issue #83214.

Ideally we should have a full fledged lexer similar to `schema_type_parser.cpp`, since regex can be more and more difficult to read if we add more support to it.

Adding this to unblock this issue for now.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[torchgen] Add logic in annotation parser to accept alias set"


Extending the current regex in `model.py` to support annotation alias set. See issue #83214.

Ideally we should have a full fledged lexer similar to `schema_type_parser.cpp`, since regex can be more and more difficult to read if we add more support to it.

Adding this to unblock this issue for now.

[ghstack-poisoned]

SherlockNoMad commented on torchgen/model.py:
```diff
@@ -1577,26 +1577,34 @@ class Annotation:
+        #   in the example.
+        # Group #6: optional sub-section of alias after set, matches "|b" in the
+        #   example.
+        m = re.match(r"^([a-z])(\|[a-z])*(!?)( -> (\*|[a-z](\|[a-z])*))?$", ann)
```
Looks like it also matches for "a|b! -> d|c" 
Is this valid annotation?

larryliu0820 replied:
"a|b!" is forbidden, see assertion in line 1604. But yes we should have more assertions to prevent stuff like "a|b -> d|c". Actually according to this [README.md](https://fburl.com/z42592t3) we should only support either one alias to alias set "a! -> a|b" or the case in `register_prim_ops.cpp`: "a|b". We shouldn't have two alias sets at the same time.

SherlockNoMad commented on torchgen/model.py:
```diff
@@ -1577,26 +1577,34 @@ class Annotation:
+        alias_set = tuple(before_alias.split("|"))
+        is_write = m.group(3) == "!"
+        assert not (
+            is_write and len(alias_set) > 1
```
I see, you already handled here...

SherlockNoMad reviewed (approved):
LGTM. Thank you!

larryliu0820 committed:
Update on "[torchgen] Add logic in annotation parser to accept alias set"


Extending the current regex in `model.py` to support annotation alias set. See issue #83214.

Ideally we should have a full fledged lexer similar to `schema_type_parser.cpp`, since regex can be more and more difficult to read if we add more support to it.

Adding this to unblock this issue for now.

[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge

larryliu0820 committed:
Update on "[torchgen] Add logic in annotation parser to accept alias set"


Extending the current regex in `model.py` to support annotation alias set. See issue #83214.

Ideally we should have a full fledged lexer similar to `schema_type_parser.cpp`, since regex can be more and more difficult to read if we add more support to it.

Adding this to unblock this issue for now.

[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge
